src/discord.rs

```diff
@@ -0,0 +1,125 @@
+#![deny(non_snake_case)]
+
+use std::num::NonZeroU32;
+
+use discord_sdk::{
+    activity::{ActivityBuilder, Assets, PartyPrivacy},
+    user::User,
+    wheel::UserState,
+    wheel::Wheel,
+    Discord, DiscordApp, Subscriptions,
+};
+use rrplug::prelude::*;
+
+use crate::exports::PLUGIN;
+
+/// the discord app's id, taken from older v1 discord rpc
+const APP_ID: i64 = 941428101429231617;
+
+/// struct to hold everything required to run discord rpc
+pub struct Client {
+    pub discord: Discord,
+    pub user: User,
+    pub wheel: Wheel,
+}
+
+/// discord rpc update function
+/// 
+///  doesn't run on the titanfall 2 thread since it needs async
+pub async fn async_main() {
+    let activity = &PLUGIN.wait().activity;
+
+    let client = match make_client(Subscriptions::ACTIVITY).await {
+        Ok(c) => c,
+        Err(_) => {
+            log::error!("Is your discord running?");
+            return;
+        }
+    };
+
+    match client.discord.clear_activity().await {
+        Ok(_) => log::info!("cleared activity"),
+        Err(err) => log::error!("coudln't clear activity because of {:?}", err),
+    }
+
+    loop {
+        let data = activity.lock().clone();
+
+        if let Some(img) = &data.large_image {
+            if img.is_empty() {
+                log::warn!("img is empty at {:?}", data.last_state);
+            }
+        }
+
+        let mut activity_builder = ActivityBuilder::default()
+            .details(data.details)
+            .state(data.state)
+            .assets(Assets {
+                large_image: data.large_image,
+                large_text: data.large_text,
+                small_image: data.small_image,
+                small_text: data.small_text,
+            });
+
+        if let Some(start) = data.start {
+            activity_builder = activity_builder.start_timestamp(if start == 0 { 1 } else { start });
+        }
+        if let Some(end) = data.end {
+            activity_builder = activity_builder.end_timestamp(end);
+        }
+        if let Some(party) = data.party {
+            activity_builder = activity_builder.party(
+                "whar",
```

Comment:
Based on the string comment I assume we would never hit this state with current Northstar?

Reply:
it's used to display the number of players in the presence; the string is just there for the non working party system discord had.